### PR TITLE
Leaky bucket

### PR DESCRIPTION
**Pull request by WJWH**
Replaces the MGET algorithm with a leaky bucket calculation in a Lua script. If the script is not yet present in the Redis script cache it will be reloaded from disk. 

This replicates the block that is instated if someone makes too many calls, so it's technically not just a throttler but also a circuit breaker of sorts.

### CODE REVIEW

File: lib/prorate/throttle.rb [hunk 1]
```diff
@@ -1,11 +1,24 @@
...
+  end
+
+  class ScriptHashMismatch < StandardError
+  end
+
+  class MisconfiguredThrottle < StandardError
+  end
+
   class Throttle < Ks.strict(:name, :limit, :period, :block_for, :redis, :logger)
+    CURRENT_SCRIPT_HASH = 'fec1e501aaf0a47b8778e46bdb7d0945527e8c0c'
```
Comment (julik) [inline lib/prorate/throttle.rb]: Shouldn't this be computed from the file at load time?

File: lib/prorate/throttle.rb [hunk 2]
```diff
@@ -17,18 +30,30 @@ def throttle!
...
-
         logger.info { "Applying throttle counter %s" % name }
-        c = Prorate::Counter.new(redis: r, id: identifier, logger: logger, window_size: period)
-        after_increment = c.incr
-        if after_increment > limit
+        resp = run_lua_throttler(redis: r, identifier: identifier, bucket_capacity: limit, leak_rate: @leak_rate, weight: 1, block_for: block_for)
+
+        if resp != "OK"
           logger.warn { "Throttle %s exceeded limit of %d at %d" % [name, limit, after_increment] }
-          Prorate::BlockFor.block!(redis: r, id: identifier, duration: block_for)
```
Comment (julik) [inline lib/prorate/throttle.rb]: If you set a block for a certain time (a blanket block) it will substantially shorten the amount of stuff you have to do in Redis once refused.

File: lib/prorate/throttle.rb [hunk 3]
```diff
@@ -17,18 +30,30 @@ def throttle!
...
-          raise Throttled.new(period)
+          raise Throttled.new("Throttled, please lower your temper and try again in #{resp} seconds")
```
Comment (julik) [inline lib/prorate/throttle.rb]: The `retry_in_seconds` on the exception that gets raised is actually important, because it's value can then be used for the `Retry-After` header value.

File: lib/prorate/throttle.rb [hunk 4]
```diff
@@ -17,18 +30,30 @@ def throttle!
...
+
+    def run_lua_throttler(redis: , identifier: , bucket_capacity: , leak_rate: , weight: , block_for: )
+      # Slightly magic hash:
+      redis.evalsha(CURRENT_SCRIPT_HASH, [], [identifier, bucket_capacity, leak_rate, weight, block_for])
+    rescue Redis::CommandError => e
+      if e.message.include? "NOSCRIPT"
+        # The Redis server has never seen this script before. Needs to run only once in the entire lifetime of the Redis server (unless the script changes)
+        script_filepath = File.join(__dir__,"rate_limit.lua")
+        script = File.read(script_filepath)
+        raise ScriptHashMismatch if Digest::SHA1.hexdigest(script) != CURRENT_SCRIPT_HASH
```
Comment (julik) [inline lib/prorate/throttle.rb]: See above re. computing this hash at-load

File: scripts/bm.rb [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+logz = Logger.new(STDERR)
+logz.level = Logger::FATAL # block out most stuff
+
+times = []
+15.times do
+  times << Benchmark.realtime {
+    t = Prorate::Throttle.new(redis: r, logger: logz, name: "throttle-login-email", limit: 60, period: 120, block_for: 5)
+    # Add all the parameters that function as a discriminator
+    t << '127.0.2.1'
+    t << 'onno@wetransfer.com'
```
Comment (julik) [inline scripts/bm.rb]: Sure you want his email in code on GH? :-)

File: scripts/reload_lua.rb [hunk 6]
```diff
@@ -0,0 +1,6 @@
+# Reloads the script into redis and prints out the SHA it can be called with
+require 'redis'
+r = Redis.new
+script = File.read('../lib/prorate/rate_limit.lua')
+sha = r.script(:load,script)
+puts sha
```
Comment (julik) [inline scripts/reload_lua.rb]: Trailing whitespace

File: scripts/bm.rb [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+  times << Benchmark.realtime {
+    t = Prorate::Throttle.new(redis: r, logger: logz, name: "throttle-login-email", limit: 20, period: 120, block_for: 5)
+    # Add all the parameters that function as a discriminator
+    t << ip
+    t << email
+    t.throttle!
+  }
+end
+
+puts average_ms times
```
Comment (julik) [inline scripts/bm.rb]: Trailing whitespace

File: spec/counter_spec.rb [hunk 8]
```diff
@@ -1,53 +0,0 @@
-require 'spec_helper'
-
-describe Prorate::Counter do
```
Comment (julik) [inline spec/counter_spec.rb]: The idea behind having a counter was also to be able to give decent info at some stage regarding the duration of the throttle and the requests used versus quota (see how the GH API operates). Granted - I didn't use that in the monolith, but it is one of the things which are really nice to have. Does the Lua setup support such a use case? (see how the GH API reports limits and throttling via headers)

File: lib/prorate/throttle.rb (hunk 1 above)
Comment (WJWH) [inline lib/prorate/throttle.rb]: It is now so.

File: lib/prorate/throttle.rb (hunk 2 above)
Comment (WJWH) [inline lib/prorate/throttle.rb]: This is now done in Lua instead of with a separate GET call to save on roundtrips.

File: lib/prorate/throttle.rb (hunk 3 above)
Comment (WJWH) [inline lib/prorate/throttle.rb]: Re-added this.

File: lib/prorate/throttle.rb (hunk 4 above)
Comment (WJWH) [inline lib/prorate/throttle.rb]: Fixed.

File: scripts/bm.rb (hunk 5 above)
Comment (WJWH) [inline scripts/bm.rb]: Changed to something else. It'll still be there for people digging through the history though :(

File: scripts/bm.rb (hunk 7 above)
Comment (WJWH) [inline scripts/bm.rb]: Fixed

File: scripts/reload_lua.rb (hunk 6 above)
Comment (WJWH) [inline scripts/reload_lua.rb]: Fixed

File: spec/counter_spec.rb (hunk 8 above)
Comment (WJWH) [inline spec/counter_spec.rb]: The `throttle!` method now returns an integer indicating how many more calls you can make before you'll hit the block.

File: scripts/bm.rb (hunk 5 above)
Comment (onnos) [inline scripts/bm.rb]: bonjour

**Comment by WJWH** [comment]
Improved the Lua script with help from @thomasvnoort; it now has better naming, more idiomatic Lua use and better error handling for an edge case. 

We also made another test script to see the difference in 'maximum time blocking redis'; both with an empty redis and with one containing 4 million keys already. It seems that the access time for GET requests is pretty constant with data set size. (Only proper, as it's a hashmap, but still) Either way, even the longest request (the one containing the MGET) still takes three times as long as the entire Lua script. The difference in speed for the `throttle!` call is also about three times.
